python/pyiceberg/io/pyarrow.py

```diff
@@ -437,3 +465,198 @@ def visit_or(self, left_result: pc.Expression, right_result: pc.Expression) -> p
 
 def expression_to_pyarrow(expr: BooleanExpression) -> pc.Expression:
     return boolean_expression_visit(expr, _ConvertToArrowExpression())
+
+
+def project_table(
+    files: Iterable[FileScanTask], table: Table, row_filter: BooleanExpression, projected_schema: Schema, case_sensitive: bool
+) -> pa.Table:
+    """Resolves the right columns based on the identifier
+
+    Args:
+        files(Iterable[FileScanTask]): A URI or a path to a local file
+        table(Table): The table that's being queried
+        row_filter(BooleanExpression): The expression for filtering rows
+        projected_schema(Schema): The output schema
+        case_sensitive(bool): Case sensitivity when looking up column names
+
+    Raises:
+        ResolveException: When an incompatible query is done
+    """
+
+    if isinstance(table.io, PyArrowFileIO):
+        scheme, path = PyArrowFileIO.parse_location(table.location())
+        fs = table.io.get_fs(scheme)
+    else:
+        raise ValueError(f"Expected PyArrowFileIO, got: {table.io}")
+
+    bound_row_filter = bind(table.schema(), row_filter, case_sensitive=case_sensitive)
+
+    projected_field_ids = {
+        id for id in projected_schema.field_ids if not isinstance(projected_schema.find_type(id), (MapType, ListType))
```

Comment:
Why does this exclude maps and lists?

Reply:
I think this is related to the use of `select_full_types=False` below, but that flag is based on the logic in the Java implementation's `GetProjectedIds`. That visitor accumulates a list of field IDs throughout the schema that are projected, although it doesn't include map or list fields because the map key/value and list element fields are directly selected.

If this is going to use `select_full_types=False` then it should apply the same visitor logic and set the visitor that returns projected IDs to include struct IDs.

Reply:
This is because we can't select the MapType/ListType:
```
                if not field.field_type.is_primitive:
>                   raise ValueError(
                        f"Cannot explicitly project List or Map types, {field.field_id}:{field.name} of type {field.field_type} was selected"
                    )
E                   ValueError: Cannot explicitly project List or Map types, 5:ids of type list<int> was selected
```
We're only interested in the element in case of the list, and the key-value in case of a map.

Reply:
Okay, so this is because of the use of `select_full_types=False`.

I think that this works since it is using `schema.field_ids` but we may want to move this to a helper method eventually, rather than embedding it here.

Reply:
The question is then if we want to return full structs.

If the requested schema is:
```
location: struct<lat: double>
```

Of the table schema:
```
location: struct<lat: double, long: double>
```

With the `select_full_types=True` it would return:
```
location: struct<lat: double, long: double>
```

Instead of just lat.